### ISSUE TITLE
`cgo_library()` fails to compile if there's an in-package symbol dependency on non-cgo file

### ISSUE THREAD

**Issue by treaster** [bazelbuild/rules_go#255]
`cgo_libary()` fails to compile if the cgo source file has a symbol dependency on a go source file in the same package. The error is `undefined: [symbol].`

More specifically:

- A package cgo_bad_lib contains two files: cgo_source.go and go_source.go.
- go_source.go defines unexported symbol goFunc().
- cgo_source.go relies on goFunc().
- go build can compile the cgo_bad_lib package, so the source code is correct.
- bazel build fails, citing undefined: goFunc. The cgo and go source files appear to be compiled as separate units.

`go build` can build all provided targets successfully.

```
$ bazel version
Build label: 0.4.3- (@non-git)
Build target: bazel-out/local-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Mon Jan 30 22:07:29 2017 (1485814049)
Build timestamp: 1485814049
Build timestamp as int: 1485814049
```

Instructions to reproduce the problem can be found at https://github.com/treaster/bazel_cgo_example, also copied below:

1. `git clone https://github.com/treaster/bazel_cgo_example.git && export GOPATH=$(pwd)/bazel_cgo_example && cd $GOPATH/src
go build example.com/cgo_bad_lib # Building with go build works`
2. `bazel build example.com/go_ok_lib/... # bazel configuration works`
3. `bazel build example.com/cgo_ok_lib/... # bazel+cgo can work`
4. `bazel build example.com/cgo_bad_lib/... # bazel+cgo fails when a cgo file depends on a pure-Go symbol`

**Comment by pmbethe09** [comment]
What do your BUILD files look like for this?

**Comment by treaster** [comment]
BUILD files are in the example repo I linked to.

Root BUILD file:
https://github.com/treaster/bazel_cgo_example/blob/master/src/BUILD

cgo library BUILD file:
https://github.com/treaster/bazel_cgo_example/blob/master/src/example.com/cgo_bad_lib/BUILD

In the latter case, I tried moving all of the source files into the cgo_library, but I get compile errors complaining that `import "C"` is missing from the non-cgo files.

**Comment by pmbethe09** [comment]
I don't know if the current cgo support handles this case.
Untested, but it might work if you remove 'library =' from the go_default_library, and instead put a 'library =' tag on the cgo_library referencing the go_library.
And then make sure the external reference to cgo_bad_lib is to :cgo_default_library

**Comment by treaster** [comment]
Thanks for the quick responses!

Current cgo definitely handles this case. I see now that there was a formatting error in my original post that obscured that instruction. 

Running `go build example.com/cgo_bad_lib` in the `src/` directory of my example repo does the right thing. Somehow, it's able to compile all files in the package as a single unit.

I followed your suggestion and I was able to get the library to build. However, I think this is avoiding the underlying problem of multiple-unit compilation. If I have interdependencies back and forth across the Go/cgo divide, this approach fails. I've updated my demo repo to demonstrate the problem. Now go -> cgo -> go. `go build` can still build this library successfully.
https://github.com/treaster/bazel_cgo_example/blob/master/src/example.com/cgo_bad_lib/go_source.go
https://github.com/treaster/bazel_cgo_example/blob/master/src/example.com/cgo_bad_lib/cgo_source.go

If I recall from my time at Google (a few years ago, so I may be misremembering...), the go rules for blaze had separate attributes for `srcs` vs. `cgo_srcs` or something like that, and there was no separate `cgo_library()` rule. I know rules_go is pretty far along, but maybe something like this is possible to compile all files for a package in the same unit?

Thanks so much for your help!

-Mike
